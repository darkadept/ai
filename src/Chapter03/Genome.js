import slice from 'lodash/slice';
import {random} from '../lib';

export default class Genome {
	constructor(numBits) {
		this._vecBits = [];
		this._fitness = 0;

		for (let i=0; i<numBits; i++) {
			this._vecBits.push(random(0, 1));
		}
	}

	get fitness() {return this._fitness};
	set fitness(value) {this._fitness = value;}

	get length() {return this._vecBits.length};
	slice(start, numBits) {
		return slice(this._vecBits, start, start+numBits);
	}
	// get bits() {return this._vecBits;}

	// equals(b) {
	// 	return this._vecBits.join('') === b._vecBits.join('');
	// }
}
