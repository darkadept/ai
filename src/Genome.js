import {random} from './lib';

export default class Genome {
	constructor(numBits) {
		this._vecBits = [];
		this._fitness = 0;

		for (let i=0; i<numBits; i++) {
			this._vecBits.push(random(0, 1));
		}
	}

	get fitness() {return this._fitness};
	set fitness(value) {this._fitness = value;}
	get vector() {return this._vecBits;}
}
