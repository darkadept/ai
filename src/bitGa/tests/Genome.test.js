import Genome from '../Genome';

const mockMath = Object.create(global.Math);
let randomFlip = true;
mockMath.random = () => {
	if (randomFlip) {
		randomFlip = false;
		return 0.3;
	}
	randomFlip = true;
	return 0.6;
};
global.Math = mockMath;

it('should construct empty', () => {
	const g = new Genome();

	expect(g.fitness).toBe(0);
	expect(g.length).toBe(0);
});

it('should construct random', () => {
	const g = new Genome({numBits: 10});

	expect(g.fitness).toBe(0);
	expect(g.length).toBe(10);
	expect(g.bits).toMatchSnapshot();
});

it('should construct with fitness', () => {
	const g = new Genome({numBits: 10, fitness: 3});

	expect(g.fitness).toBe(3);
	expect(g.length).toBe(10);
});

it('should construct as a clone', () => {
	const g = new Genome({numBits: 10});
	const g2 = new Genome({bits: g.bits});

	expect(g2.length).toBe(10);
	expect(g2.bits).toMatchSnapshot();
});

it('should clone & push bits', () => {
	const g = new Genome({numBits: 10});
	const g2 = g.clone();
	g2.push(1);

	expect(g.length).toBe(10);
	expect(g).toMatchSnapshot();
	expect(g2.length).toBe(11);
	expect(g2).toMatchSnapshot();
});

it('should flip bits', () => {
	const g = new Genome({numBits: 10});
	g.flip(0);
	g.flip(2);
	g.flip(4);
	g.flip(6);
	g.flip(8);
	expect(g).toMatchSnapshot();
});

it('should slice', () => {
	const a = new Genome({numBits: 8});
	const b = a.slice(0, 4);
	const c = a.slice(4, 4);

	expect(b).toMatchSnapshot();
	expect(c).toMatchSnapshot();
});
