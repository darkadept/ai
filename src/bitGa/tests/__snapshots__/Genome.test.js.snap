// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should clone & push bits 1`] = `
Genome {
  "_bits": Array [
    0,
    1,
    0,
    1,
    0,
    1,
    0,
    1,
    0,
    1,
  ],
  "_fitness": 0,
}
`;

exports[`should clone & push bits 2`] = `
Genome {
  "_bits": Array [
    0,
    1,
    0,
    1,
    0,
    1,
    0,
    1,
    0,
    1,
    1,
  ],
  "_fitness": 0,
}
`;

exports[`should construct as a clone 1`] = `
Array [
  0,
  1,
  0,
  1,
  0,
  1,
  0,
  1,
  0,
  1,
]
`;

exports[`should construct random 1`] = `
Array [
  0,
  1,
  0,
  1,
  0,
  1,
  0,
  1,
  0,
  1,
]
`;

exports[`should flip bits 1`] = `
Genome {
  "_bits": Array [
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
  ],
  "_fitness": 0,
}
`;

exports[`should slice 1`] = `
Array [
  0,
  1,
  0,
  1,
]
`;

exports[`should slice 2`] = `
Array [
  0,
  1,
  0,
  1,
]
`;
